### PR TITLE
Rewrite container docs for custom containers

### PR DESCRIPTION
**Pull request by emilymye**
This rewrites the container guide as we move to launch custom containers for Dataflow runner. There are two methods that should be highlighted - one is using the released Dockerhub images as a base image, and one is building from Beam source itself. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by emilymye** [comment]
R: @tysonjh 
R: @tvalentyn 
R: @alanmyrvold
R: @rosetn

Current page: https://beam.apache.org/documentation/runtime/environments/

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (tysonjh) [inline website/www/site/content/en/documentation/runtime/environments.md:218]: Please put this into a code block or reformat.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (tvalentyn) [reply website/www/site/content/en/documentation/runtime/environments.md]: s/existing prebuilt/released ?

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (tvalentyn) [inline website/www/site/content/en/documentation/runtime/environments.md:36]: This is also a soft requirement - users could use CloudBuild or other equivalents that can build a container on demand from a given Dockerfile.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (emilymye) [reply website/www/site/content/en/documentation/runtime/environments.md:224]: The {{< highlight class="runner-X" >}} formats it into code-blocks, tabbed by runner type - see existing page https://beam.apache.org/documentation/runtime/environments/#testing-customized-images

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (tvalentyn) [reply website/www/site/content/en/documentation/runtime/environments.md:163]: I think in the future there is a plan for Portable runner to support Dataflow endpoint, not sure how far on the roadmap that is.

**Comment by tvalentyn** [comment]
(some of my comments got hidden under `Show Resolved` due to new push, hope you can still find them :) )

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (tvalentyn) [inline website/www/site/content/en/documentation/runtime/environments.md:30]: I meant adding only one of these, no strong opinion which wording is better.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (tvalentyn) [inline website/www/site/content/en/documentation/runtime/environments.md:142]: Re preexisting documentation: `With these settings, the
docker command-line tool will implicitly try to push the container to the Docker Hub Apache repository.`

Docker will not implicitly push images unless instructed to. Consider removing this phrase.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (tvalentyn) [inline website/www/site/content/en/documentation/runtime/environments.md:94]: I would recommend users to start their customization from a stable, a released version of a Dockerfile (and later use the SDK of the same version when they launch their pipeline), so consider:
1. instead of modifying with code at master, switch to a release branch (e.g. `git checkout origin/release-2.26.0`).
2. instead of using `latest` as tag for custom image in the snippet below, use something like 2.26.0_custom.

**Review by tvalentyn — APPROVED** [review]
LGTM, feel free to ping me when this is ready to merge.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (tysonjh) [reply website/www/site/content/en/documentation/runtime/environments.md:223]: Got it. I went to 'view file' in GitHub and it only showed the rendered markdown which made each of these H1 titles. Is there a way to preview?

**Review by tysonjh — APPROVED** [review]

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (ibzib) [reply website/www/site/content/en/documentation/runtime/environments.md:229]: The best way is to serve the website from source: `./gradlew :website:serveWebsite`

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (ibzib) [reply website/www/site/content/en/documentation/runtime/environments.md:171]: @tvalentyn Can you send me more information on this plan if you have any?

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (ibzib) [inline website/www/site/content/en/documentation/runtime/environments.md:170]: Starting with Beam 2.25.0, prefer using `--environment_options=docker_container_image=foo` instead of `--environment_config`, since `--environment_config` causes confusion when used with different (non-Docker) environments (BEAM-10671).

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (ibzib) [inline website/www/site/content/en/documentation/runtime/environments.md:189]: It'd be nice to use `FlinkRunner` instead, which automatically starts a job server so the user doesn't have to. But it doesn't have to block this PR, I can do it in a later change.

Comment (emilymye) [reply website/www/site/content/en/documentation/runtime/environments.md:189]: We are probably waiting until early next week to merge in - probably better to use the more accurate version of commands. Do you have an example of recommended arguments with FlinkRunner for Python/is that supported? I couldn't find one in the docs. 

Also, should we be doing the same things for spark? Based on a quick look it doesn't look like we have the same functionality for Spark, but I'm not very familiar with either Flink or Spark.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (emilymye) [reply website/www/site/content/en/documentation/runtime/environments.md:170]: Do you think we need to note this? Or should we just recommend `--environment_options`. If it was only supported post 2.25.0, we can maybe say "Older versions of the SDK (<2.25.0) may not support `--environment_options`. For these versions, use `--environment_config=$IMAGE_URL` instead?"

Comment (ibzib) [reply website/www/site/content/en/documentation/runtime/environments.md:170]: Since 2.25.0 is still pretty recent, maybe we should keep this as it is for now. Perhaps we can add a TODO to update this later?

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (ibzib) [reply website/www/site/content/en/documentation/runtime/environments.md:189]: It'd look something like this:

```sh
python -m apache_beam.examples.wordcount \
--input=/path/to/inputfile \
--output=/path/to/write/counts \
--runner=FlinkRunner \
--environment_type=DOCKER \
--environment_config="${IMAGE}:${TAG}"
```

And then the same for Spark, but substitute `--runner=SparkRunner`.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (emilymye) [reply website/www/site/content/en/documentation/runtime/environments.md:170]: added a TODO; I think it makes sense to update this all together, maybe when we move to GA for custom containers or port options over to DataflowRunner.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (emilymye) [reply website/www/site/content/en/documentation/runtime/environments.md:189]: edited - I took out the Flink/Spark runner server comments and code.

**Review by ibzib — APPROVED** [review]
LGTM.

It's exciting to finally have custom containers on Dataflow publicly documented.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (tysonjh) [reply website/www/site/content/en/documentation/runtime/environments.md:173]: +1 I'd like to know about this as well since we shouldn't go to GA with custom containers until the flags are settled.

Comment (tvalentyn) [reply website/www/site/content/en/documentation/runtime/environments.md:173]: @ananvay @angoenka @robertwb may have input on this plan.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (rosetn) [reply website/www/site/content/en/documentation/runtime/environments.md:236]: You can also view the details of Website_Stage_GCS ("Run Website_Stage_GCS PreCommit") to find the staged version of the website. You can rerun the test if the link's gone stale.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (rosetn) [inline website/www/site/content/en/documentation/runtime/environments.md:52]: Delete this line "Steps:"

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (rosetn) [inline website/www/site/content/en/documentation/runtime/environments.md:54]: Can you combine "As an example, this `Dockerfile`:" to the sentence after the code sample? It's a little confusing to have the interrupted sentence.

There are a few more instances of sentences with a code snippets in the Modifying a source Dockerfile in Beam
section--can you combine those too? The typical pattern is instruction>code snippet>explanation.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (rosetn) [inline website/www/site/content/en/documentation/runtime/environments.md:82]: If it's important to check the image IDs and digest, let's make this note into its own step. Same with the other instances of this note. Readers will sometimes ignore notes.

"3. After pushing a container image, you should verify the remote image ID and digest should match the local image ID and digest, output from `docker build` or `docker images`."

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (rosetn) [inline website/www/site/content/en/documentation/runtime/environments.md:102]: Does "Navigate to the root directory where you've installed your local copy of the Beam SDK." work?

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (rosetn) [inline website/www/site/content/en/documentation/runtime/environments.md:146]: I think it's OK to keep this at H3 instead of H4

### Additional build parameters

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (rosetn) [inline website/www/site/content/en/documentation/runtime/environments.md:133]: I'd remove "(see below)" and either explicitly link to that heading or use "see the following section."

**Review by rosetn — CHANGES_REQUESTED** [review]
Thanks for rewriting this!

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (emilymye) [reply website/www/site/content/en/documentation/runtime/environments.md]: Tried to make it clearer (added some env var setting in instructions) - mostly worried installed might get confused for package installation (i.e. where pip installed Beam)

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (emilymye) [reply website/www/site/content/en/documentation/runtime/environments.md:88]: hmm this was working on my local build of the website, will need to figure out the correct URL

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (ananvay) [reply website/www/site/content/en/documentation/runtime/environments.md:178]: For Dataflow, I thought the idea was to use --sdk_harness_container_image or something like this? I'm not sure I have the whole context off the top of my head here, but I'd be happy to chat! :)

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (emilymye) [reply website/www/site/content/en/documentation/runtime/environments.md]: Figured it out!

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (tysonjh) [inline website/www/site/content/en/documentation/runtime/environments.md:173]: Can you add an anchor to this so I can link to it in cloud docs please?

Comment (emilymye) [reply website/www/site/content/en/documentation/runtime/environments.md]: added #{running-pipelines}

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (rosetn) [inline website/www/site/content/en/documentation/runtime/environments.md:256]: Typo here--I also think you can just call this section "Troubleshooting" or "Considerations"

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (rosetn) [inline website/www/site/content/en/documentation/runtime/environments.md:258]: Can you write an introductory sentence for this list?

**Review by rosetn — CHANGES_REQUESTED** [review]

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (robertwb) [reply website/www/site/content/en/documentation/runtime/environments.md:178]: Regardless of the plans for using the Portability API for submitting jobs, we should standardize on a single flag for specifying the image across runners. It could be argued that environment_config is not the best flag for this; maybe we should take it to the list?

Comment (ibzib) [reply website/www/site/content/en/documentation/runtime/environments.md:178]: > Regardless of the plans for using the Portability API for submitting jobs, we should standardize on a single flag for specifying the image across runners. It could be argued that environment_config is not the best flag for this; maybe we should take it to the list?

+1 I agree that standardizing the flag is a worthwhile improvement, but it's out of scope for this PR.

Comment (robertwb) [reply website/www/site/content/en/documentation/runtime/environments.md:178]: Standardizing the flag should be done before we document and adveritize this feature.

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (rosetn) [inline website/www/site/content/en/documentation/runtime/environments.md:61]: Extra colon after `Dockerfile`

**Review by rosetn — APPROVED** [review]
One small typo but otherwise LGTM

File: website/www/site/content/en/documentation/runtime/environments.md
Comment (emilymye) [reply website/www/site/content/en/documentation/runtime/environments.md:177]: Our plan was to keep these as is for preview and standardize for GA. This is also editing existing documentation which used --environment_config. I can send an email to dev re:flags.

Comment (robertwb) [reply website/www/site/content/en/documentation/runtime/environments.md:177]: If we're supporting old SDKs, we have to support the old flag anyway, so we should go ahead and get this in and standardizing going forward.

**Pull request merged**
